students/NatalieRodriguez/Lesson05/mailroom3.py

```diff
@@ -0,0 +1,169 @@
+
+
+def donor_dashboard():
+    # dict of menu options
+    options = {'1': ("Send a thank you note.", thank_you),
+               '2': ("Create a report.", create_report),
+               '3': ("Send thank yous to all donors.", thank_all),
+               '4': ("Exit the donor dashboard.", exit)}
+
+    response = ''
+    while response != '4':
+        print()
+        for i in options:
+            print(i, options[i][0])
+
+        # Get user's selection
+        response = ''
+        while not response in options:
+            response = input("Type your selection: ").strip()
+
+        options[response][1]()  # Call helper function
+
+
+def exit():
+
+    print("\nExiting the donor dashboard. Goodbye! \n")
+    return
+
+
+def thank_you():
+    """
+    Add new donations for new or existing donors, and send a thank-you
+    letter.
+    :return:  None.
+    """
+    # Get the donor name, show all donors, or quit
+    response = input(
+        "\nType the full donor name (Enter 'list' to show all donors or enter 'exit'): "
+    ).strip()
+
+    if response.lower() in ('', 'exit'):
+        exit()
+        return
+
+    elif response.lower() == 'list':
+        print("\nDONORS:")
+        for donor in donors:
+            print(donor)
+        thank_you()
+
+    else:
+        while True:
+            donation = input(
+                f"Enter amount of donation by '{response}' (or enter 'exit'): "
+            ).strip().lower()
+            if donation == 'exit':
+                exit()
+                return
+
+            elif donation.strip('0123456789.') == '' and len(
+                    donation) > 0 and donation.count('.') <= 1 and float(
+                donation) > 0.0:
+                break
+
+        donation = float(donation)
+        donors.setdefault(response, [])
+        donors[response].append(donation)
+        print(create_form_letter(response, donation))
+
+
+def create_report():
+
+    print('\n')
+    print('        Donor Name        |    Total Given      | '
+          + 'Number of Gifts |    Average Gift    ')
+    print('--------------------------|---------------------|-'
+          + '----------------|--------------------')
+    for individual_donor, donations in donors.items():
+        total_donation = sum(donations)
+        number_of_gifts = len(donations)
+        average_donation = 1.0 * total_donation / number_of_gifts
+        print('{:<25s} | ${:>18,.2f} | {:>15d} | ${:>18,.2f}'.format(
+            individual_donor, total_donation,
+            number_of_gifts, average_donation))
+
+
+def thank_all():
+
+    cur_dir = os.getcwd()
+    print('\nThe current directory is %s' % cur_dir)
+    new_dir = input('\nEnter the directory you want to save to '
+                    '(Press enter to save in the current location): ').strip()
+#exception handling changes here: indented all for try/except block
+    try:
```

Comment:
It's a good idea to try and have as little code inside your `try` block as possible, so that you only include code that might raise the expected exception. Otherwise, you can end up with unanticipated and unhandled errors slipping by.